In [2]:
import pandas as pd
import matplotlib.pyplot as plt
from sklearn import tree
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier

In [3]:
df = pd.read_excel('dataset/Bank_Personal_Loan_Modelling.xlsx',sheet_name=1)

In [4]:
df.head()

ID  Age  Experience  Income  ZIP Code  Family  CCAvg  Education  Mortgage  \
0   1   25           1      49     91107       4    1.6          1         0   
1   2   45          19      34     90089       3    1.5          1         0   
2   3   39          15      11     94720       1    1.0          1         0   
3   4   35           9     100     94112       1    2.7          2         0   
4   5   35           8      45     91330       4    1.0          2         0   

   Personal Loan  Securities Account  CD Account  Online  CreditCard  
0              0                   1           0       0           0  
1              0                   1           0       0           0  
2              0                   0           0       0           0  
3              0                   0           0       0           0  
4              0                   0           0       0           1

In [5]:
df.shape

(5000, 14)

In [10]:
df.columns

Index(['ID', 'Age', 'Experience', 'Income', 'ZIP Code', 'Family', 'CCAvg',
       'Education', 'Mortgage', 'Personal Loan', 'Securities Account',
       'CD Account', 'Online', 'CreditCard'],
      dtype='object')

In [11]:
df.drop(['ID'],axis=1,inplace=True)

In [12]:
df.isnull().sum()

Age                   0
Experience            0
Income                0
ZIP Code              0
Family                0
CCAvg                 0
Education             0
Mortgage              0
Personal Loan         0
Securities Account    0
CD Account            0
Online                0
CreditCard            0
dtype: int64

In [13]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 13 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Age                 5000 non-null   int64  
 1   Experience          5000 non-null   int64  
 2   Income              5000 non-null   int64  
 3   ZIP Code            5000 non-null   int64  
 4   Family              5000 non-null   int64  
 5   CCAvg               5000 non-null   float64
 6   Education           5000 non-null   int64  
 7   Mortgage            5000 non-null   int64  
 8   Personal Loan       5000 non-null   int64  
 9   Securities Account  5000 non-null   int64  
 10  CD Account          5000 non-null   int64  
 11  Online              5000 non-null   int64  
 12  CreditCard          5000 non-null   int64  
dtypes: float64(1), int64(12)
memory usage: 507.9 KB


In [14]:
randomForestClassifier = RandomForestClassifier(n_estimators=1000,max_features=12,oob_score=True)

In [16]:
randomForestClassifier.fit(df.drop(['Personal Loan'],axis=1),df['Personal Loan'])

RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
                       max_depth=None, max_features=12, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=1000,
                       n_jobs=None, oob_score=True, random_state=None,
                       verbose=0, warm_start=False)

In [19]:
z = zip(df.columns,randomForestClassifier.feature_importances_)
z_sorted = sorted(z,key= lambda x : x[1])
for feature,importance in z_sorted:
    print(feature,'-',importance)

Personal Loan - 0.0014162211716866357
Online - 0.0021724568062479045
CD Account - 0.0037855996662823797
Mortgage - 0.006116762230448804
Experience - 0.014290995214683971
Age - 0.016807293909644763
Securities Account - 0.017464875826329815
ZIP Code - 0.01883832137170919
CCAvg - 0.06113928937546447
Family - 0.1511912696394559
Income - 0.3390535276599535
Education - 0.3677233871280927


# As we can see only 3 column is importance Family, Income, Education
* Family - 0.1511912696394559
* Income - 0.3390535276599535
* Education - 0.3677233871280927

In [20]:
decisionTreeClassifier = DecisionTreeClassifier(max_depth=4)

In [21]:
decisionTreeClassifier.fit(df[['Family','Income','Education']],df['Personal Loan'])

DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=4,
                       max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort=False,
                       random_state=None, splitter='best')

In [24]:
plt.rcParams["figure.figsize"] = (300,100)

In [ ]:
tree.plot_tree(decisionTreeClassifier,filled = True)

[Text(9825.652173913044, 4892.4, 'X[1] <= 113.5\ngini = 0.174\nsamples = 5000\nvalue = [4520, 480]'),
 Text(5822.608695652174, 3805.2, 'X[1] <= 92.5\ngini = 0.041\nsamples = 4021\nvalue = [3937, 84]'),
 Text(2911.304347826087, 2718.0, 'X[1] <= 82.5\ngini = 0.014\nsamples = 3638\nvalue = [3613, 25]'),
 Text(1455.6521739130435, 1630.7999999999997, 'X[1] <= 63.5\ngini = 0.005\nsamples = 3272\nvalue = [3263, 9]'),
 Text(727.8260869565217, 543.5999999999995, 'gini = 0.001\nsamples = 2490\nvalue = [2489, 1]'),
 Text(2183.478260869565, 543.5999999999995, 'gini = 0.02\nsamples = 782\nvalue = [774, 8]'),
 Text(4366.95652173913, 1630.7999999999997, 'X[1] <= 83.5\ngini = 0.084\nsamples = 366\nvalue = [350, 16]'),
 Text(3639.130434782609, 543.5999999999995, 'gini = 0.171\nsamples = 74\nvalue = [67, 7]'),
 Text(5094.782608695652, 543.5999999999995, 'gini = 0.06\nsamples = 292\nvalue = [283, 9]'),
 Text(8733.91304347826, 2718.0, 'X[2] <= 1.5\ngini = 0.261\nsamples = 383\nvalue = [324, 59]'),
 Text(7